doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,191 @@
+[clusters](../platform/ux.md#cluster) to Materialize.
+
+## Design
+
+### Cluster SQL syntax and semantics
+
+Four new SQL statements will be introduced with the following grammar to
+create, alter, drop, and list clusters, respectively:
+
+```
+create_cluster_stmt ::=
+  CREATE CLUSTER [IF NOT EXISTS] <name> [[WITH] <option> [, <option>...]]
+
+alter_cluster_stmt ::=
+  ALTER CLUSTER <name> [[SET] <option> [, <option>...]]
+
+drop_cluster_stmt ::=
+  DROP CLUSTER [IF EXISTS] <name> [CASCADE]
+
+show_cluster_stmt: SHOW CLUSTERS [{ LIKE 'pattern' | WHERE <expr> }]
+
+<option> ::=
+      VIRTUAL
+    | SIZE '<size>'
+
+<name> ::= <unqualified-identifier>
+```
+
+**All of these statements will be available in experimental mode only.**
+
+The name of a cluster is an unqualified identifier. Clusters are scoped to
+a catalog, and so do not belong in the usual namespace of databases and schemas.
+For a similarly scoped object in PostgreSQL, see [publications].
+
+The `CREATE CLUSTER` statement creates a new cluster. It returns an error if
+a cluster with the specified name already exists unless the `IF NOT EXISTS`
+clause is present.
+
+The `ALTER CLUSTER` statement changes properties of an existing cluster.
+
+The `DROP CLUSTER` statement drops an existing cluster. It returns an error
+if a cluster with the specified name does not exist unless the `IF EXIST`
+clause is present. Dropping a non-empty cluster (i.e., a cluster with at least
+one index) is not permitted unless `CASCADE` is present.
+
+THe `SHOW CLUSTERS` statement lists the clusters in the system, optionally
+filtered by the provided `LIKE` pattern or `WHERE` expression, which work
+analogously to the same clauses in the `SHOW DATABASES` statement.
+
+`CREATE CLUSTER` and `ALTER CLUSTER` allow specifying the same set of cluster
+options. Initially, there are only two options:
+
+  * `VIRTUAL` requests a virtual cluster. A virtual cluster may or may not
+    correspond to actual physical resources.
+
+    We suspect that the concept of a virtual cluster is primarily useful as a
+    way to exercise the cluster logic before we can allocate true clusters, and
+    it is likely that the `VIRTUAL` option is never stabilized. But they may
+    also prove to be a useful user-facing option, in which case we can consider
+    stabilization.
+
+  * `SIZE '<size>'` specifies a cluster of a fixed size. The size is an opaque
+    string from the perspective of `materialized`, but will be validated by the
+    cloud orchestration backend.
+
+    In the initial implementation, specifying the `SIZE` option will result in
+    an error during planning (i.e., not during parsing). The cloud team will be
+    along shortly to wire up this option to turn on actual physical clusters.
+
+Specifying both `VIRTUAL` and `SIZE` options in the same `CREATE CLUSTER` or
+`ALTER CLUSTER` is an error. Specifying the same option more than once is also
+an error.
+
+### Index SQL syntax and semantics
```

Comment:
These are good, but I think there will also be "use" of the `CLUSTER` argument by e.g. `SELECT`, `CREATE SINK`, `TAIL`, each of which eventually need to run in a specific compute instance. Should we start with "there is no explicit grammar for that; please use `SET CLUSTER` to indicate which cluster to use"? Which seems fine tbh, and avoids over-doing it on work until we see more. I could imagine that as `coord.rs` starts to reflect cluster identifiers more, there would be other feedback about commands that need to pick up the `CLUSTER` variable to be execute.

Reply:
Ah, good point. We _do_ need `CREATE SINK` because the catalog needs to inject the `IN CLUSTER <name>` clause into every statement that doesn't say it explicitly; that's how it remembers in which cluster each index/sink belongs when it restarts. I think we should leave `SELECT` and `TAIL` alone for now though, yeah; extending `SELECT`'s grammar is always annoying, and then for symmetry we might as well leave `TAIL` alone. Then the rule is "DDL supports explicit cluster specs; DML does not; we may add it to DML later when we've got more time to think about the right way to stuff the `IN CLUSTER` clause into `SELECT`."